### PR TITLE
RUN: Support `Run Anything` action for Cargo

### PR DESCRIPTION
**Pull request by mchernyavsky**
Closes https://github.com/intellij-rust/intellij-rust/issues/2872.

Examples:

Example1

Example1

Example1

### CODE REVIEW

**Comment by Undin** [comment]
@mchernyavsky Let's add some tests for it. Gradle [tests](https://github.com/JetBrains/intellij-community/blob/master/plugins/gradle/testSources/org/jetbrains/plugins/gradle/execution/GradleRunAnythingProviderTest.java) as example

**Comment by mchernyavsky** [comment]
Done.

File: src/183/test/kotlin/org/rust/ide/runAnything/CargoRunAnythingProviderTest.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        assertEmpty(provider.getValues(dataContext, ""))
+    }
+
+    fun `test command completion`() {
+        val commands = listOf(
+            "build", "check", "clean", "doc", "run", "test", "bench", "update", "search", "publish", "install"
+        ).map { "cargo $it" }
+        assertSameElements(provider.getValues(dataContext, "cargo"), commands)
+        assertSameElements(provider.getValues(dataContext, "cargo "), commands)
+        assertSameElements(provider.getValues(dataContext, "cargo ru"), emptyList())
```
Comment (Undin) [inline src/183/test/kotlin/org/rust/ide/runAnything/CargoRunAnythingProviderTest.kt]: Are you sure that it's desired behavior?
Personally, I expected at least `cargo run` in completion list for such input or even several `cargo run` with some args

File: src/183/test/kotlin/org/rust/ide/runAnything/CargoRunAnythingProviderTest.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import org.rust.ProjectDescriptor
+import org.rust.RsTestBase
+
+class CargoRunAnythingProviderTest : RsTestBase() {
+    private lateinit var provider: CargoRunAnythingProvider
+    private lateinit var dataContext: DataContext
+
+    override fun setUp() {
+        super.setUp()
+        provider = CargoRunAnythingProvider()
```
Comment (Undin) [inline src/183/test/kotlin/org/rust/ide/runAnything/CargoRunAnythingProviderTest.kt]: Looks like it doesn't have any state and can be initialized while `CargoRunAnythingProviderTest` object creation

File: src/183/main/kotlin/org/rust/ide/runAnything/CargoRunAnythingProvider.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    override fun getHelpCommandPlaceholder(): String = "cargo <subcommand> <args...>"
+
+    override fun getHelpCommand() = "cargo"
+
+    override fun getHelpIcon(): Icon = CargoIcons.ICON
+
+    override fun getHelpDescription(): String = "Runs Cargo command"
+
+    companion object {
+        private fun getAppropriateCargoProject(dataContext: DataContext): CargoProject? {
```
Comment (Undin) [inline src/183/main/kotlin/org/rust/ide/runAnything/CargoRunAnythingProvider.kt]: It's very similar to `org.rust.cargo.runconfig.command.RunCargoCommandActionBase#getAppropriateCargoProject(com.intellij.openapi.actionSystem.AnActionEvent)`
Let's extract common part of methods into one place

File: src/183/main/kotlin/org/rust/ide/runAnything/RunAnythingCargoItem.kt [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+                StringUtil.shortenTextWithEllipsis(" $description.", 200, 0),
+                SimpleTextAttributes.GRAYED_ITALIC_ATTRIBUTES
+            )
+        }
+
+        return component
+    }
+
+    companion object {
+        val commandDescriptions: Map<String, String> =
```
Comment (Undin) [inline src/183/main/kotlin/org/rust/ide/runAnything/RunAnythingCargoItem.kt]: Can be private

File: src/183/main/kotlin/org/rust/ide/runAnything/RsRunAnythingHelpGroup.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+package org.rust.ide.runAnything
+
+import com.intellij.ide.actions.runAnything.activity.RunAnythingProvider
+import com.intellij.ide.actions.runAnything.groups.RunAnythingHelpGroup
+
+class RsRunAnythingHelpGroup: RunAnythingHelpGroup<RunAnythingProvider<*>>() {
+    override fun getProviders(): Collection<RunAnythingProvider<*>> =
+        RunAnythingProvider.EP_NAME.extensions.filterIsInstance<CargoRunAnythingProvider>()
+
+    override fun getTitle(): String = "Rust"
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply src/183/main/kotlin/org/rust/ide/runAnything/RsRunAnythingHelpGroup.kt]: Done.

**Comment by Undin** [comment]
Perfect!
bors r+

**Pull request merged**